File: sdk/gsusb-csharp/examples/GsUsb.Examples/Examples/10_multi_custom_timing_asyncio.cs
using GsUsb;

namespace GsUsb.Examples;

internal static class Example10MultiCustomTimingAsyncio
{
    public static int Run(string[] args)
    {
        var options = CommonOptions.Parse(args, "4-channel custom timing async example");
        var (nominal, data) = DemoCommon.GetDemoCustomTimings();
        var channels = new[] { 0, 1, 2, 3 };
        Console.WriteLine($"Pair topology: {DemoCommon.BuildPairTopologyText(channels)}");
        return DemoCommon.RunMultiSummaryDemo(
            $"Running 4-channel custom timing async demo for {options.DurationSeconds:F6}s...",
            "4-channel custom timing async summary",
            channels,
            options,
            WorkerStyle.Async,
            channelIndex => new GsUsbBus(DemoCommon.MakeBusOptions(options, channelIndex, true, nominal, data)),
            (channelIndex, counter) => new CanMessage
            {
                ArbitrationId = 0x800u + (uint)channelIndex,
                IsFd = true,
                BitrateSwitch = true,
                Data = DemoCommon.MakePayload([(byte)channelIndex, (byte)(counter & 0xFF), (byte)((counter >> 8) & 0xFF)], 20, (byte)channelIndex),
            });
    }
}

File: sdk/gsusb-rs/examples/10_multi_custom_timing_asyncio.rs
#[path = "common/mod.rs"]
mod common;

use common::{build_pair_topology_text, get_demo_custom_timings, make_bus_options, make_payload, parse_common_options, run_multi_summary_demo, WorkerStyle};
use gsusb_rs::{CanMessage, GsUsbBus, Result};
use std::env;

fn main() -> Result<()> {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let options = parse_common_options(&args, "4-channel custom timing async example")?;
    let (nominal, data) = get_demo_custom_timings();
    let channels = vec![0u8, 1, 2, 3];
    println!("Pair topology: {}", build_pair_topology_text(&channels));
    run_multi_summary_demo(
        &format!("Running 4-channel custom timing async demo for {:.6}s...", options.duration_s),
        "4-channel custom timing async summary",
        &channels,
        &options,
        WorkerStyle::Async,
        |channel| GsUsbBus::new(make_bus_options(&options, channel, true, Some(nominal), Some(data))),
        |channel, counter| CanMessage {
            arbitration_id: 0x800 + u32::from(channel),
            is_fd: true,
            bitrate_switch: true,
            data: make_payload(&[channel, (counter & 0xFF) as u8, ((counter >> 8) & 0xFF) as u8], 20, channel),
            ..CanMessage::default()
        },
    )?;
    Ok(())
}